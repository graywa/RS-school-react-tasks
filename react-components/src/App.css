.app {
  min-height: 100vh;
  margin: 0 auto;  
}
.container {
  max-width: 1100px;
  width: 100%;
  height: 100%;
  margin: auto;
  padding: 0 25px;
}