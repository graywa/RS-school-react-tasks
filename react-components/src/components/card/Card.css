.card {
  position: relative;
  width: 280px;
  border-radius: 10px;
  box-shadow: 0 0 12px 2px #ebebeb;
  transition: .2s linear;
}
.card:hover {
  box-shadow: 0 0 12px 2px #c0c0c0;
}
.card__photo img {
  margin: 0 auto;
  cursor: pointer;
}
.card__discount {
  position: absolute;
  right: -15px;
  top: -10px;
  z-index: 4;
  display: flex;
  justify-content: center;
  align-items: center;
  box-shadow: 0 0 12px 2px #ebebeb;
  width: 40px;
  height: 40px;  

  background: rgb(211, 53, 53);
  border-radius: 50%;

  font-size: 12px;
  color: #fff;
}
.card__info { 
  display: flex;
  flex-direction: column;
  justify-content: end;
  height: 220px;
  padding: 0 20px 20px;
}
.card__video {
  display: flex;
  justify-content: end;
  align-items: center;
  gap: 5px;
  margin-bottom: 10px;

  color:rgb(52, 72, 161);
  font-size: 14px;
  cursor: pointer;
}
.card__video img {
  display: inline-block;
}
.card__rating-code {
  display: flex;
  justify-content: space-between;
  align-items: center;
  margin: 0 0 20px;
}
.card__rating {
  display: flex;
  align-items: end;
  gap: 5px;
  color: rgb(237 138 25);
  cursor: pointer;
}
.card__rating img {
  display: inline-block;
}
.card__code {
  font-size: 14px;
}
.card__title {
  font-size: 18px;
  font-weight: 600;

  cursor: pointer;
  transition: .2s linear;
}
.card__title:hover {
  color:rgb(211, 53, 53)
}
.card__price {
  margin-bottom: 10px;
  margin-top: auto;
  font-size: 22px;
  font-weight: 600;
}
.card__price span {
  margin-right: 5px;

  text-decoration: line-through;
  font-size: 18px;
  font-weight: 600;
  opacity: .6;
}
.card__btns {
  display: flex;
  justify-content: space-between;
  align-items: center;  
}
.card__btns button {
  border: 1px solid rgba(211, 63, 63, 0);
  border-radius: 20px;
   
  
  cursor: pointer;
  transition: .2s linear;
}
.card__buy {
  padding: 10px 40px; 
  background: rgb(211, 63, 63);

  color: #fff;
  font-size: 18px;
}
.card__buy:hover {
  background: rgb(184, 25, 25);
}
.card__cart {
  padding: 10px 20px;
  background: none;
  
}
.card__cart:hover {
  border: 1px solid rgb(211, 63, 63);
}
.card__add {
  display: flex;
  justify-content: space-between;
  padding: 10px 25px 15px;
  border-top: 1px solid rgb(214, 214, 214);
  
}
.card__add img{
  opacity: .6;
  transition: .2s linear;
  cursor: pointer;
}
.card__add img:hover {
  opacity: 1;
}