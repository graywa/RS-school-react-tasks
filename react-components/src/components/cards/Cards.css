.cards {
  display: flex;
  justify-content: space-evenly;
  flex-wrap: wrap;
  gap: 50px;
  margin: 80px 0;
}