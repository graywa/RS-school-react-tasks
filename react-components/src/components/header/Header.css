.header {
  box-shadow: 0px 0px 10px 0px #d8d8d8;
  margin-bottom: 50px;
}
.header__content {
  display: flex;
  justify-content: space-between;
  align-items: center;  
  padding: 15px 0;
}
a.active {
  font-weight: 600;
}
a:hover {
  text-decoration: underline;
}