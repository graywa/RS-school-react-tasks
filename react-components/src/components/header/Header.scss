.header {
  box-shadow: 0px 0px 12px 2px #d8d8d8;
  width: 100%;
  padding: 5px 0;
  position: fixed;
  background: #f7f7f7;
  z-index: 5;

  &__location {
    font-size: 14px;
    padding-bottom: 5px;
    
    span {
      font-weight: 600;
    }
  }
}
.header__content {
  display: flex;
  justify-content: space-between;
  align-items: center;
  padding: 20px 0;

  font-size: 20px;

  a.active {
    font-weight: 600;
  }
  a:hover {
    text-decoration: underline;
  }
}

.links {
  display: flex;
  gap: 20px;
}
