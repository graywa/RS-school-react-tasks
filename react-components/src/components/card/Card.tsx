import React from 'react';
import './Card.css';
import play from './assets/play.svg';
import star from './assets/star.svg';
import cart from './assets/cart.svg';
import heart from './assets/heart.svg';
import bell from './assets/bell.svg';

interface IProps {
  photo: string;
  video: boolean;
  rating: number;
  code: number;
  title: string;
  price: number;
  discount: number | null;
}

class Card extends React.Component<IProps> {
  render() {
    const { discount, photo, code, price, rating, title, video } = this.props;
    return (
      <div className="card">
        {discount && <div className="card__discount">-{discount}%</div>}
        <div className="card__photo">
          <img width={180} src={photo} alt="photo" />
        </div>
        <div className="card__info">
          {video && (
            <div className="card__video">
              <img width={14} src={play} alt="play" />
              <span>Видео</span>
            </div>
          )}
          <div className="card__rating-code">
            <div className="card__rating">
              <img width={20} src={star} alt="star" />
              <span>{rating}</span>
            </div>
            <div className="card__code">
              Код: <span>{code}</span>
            </div>
          </div>
          <div className="card__title">{title}</div>
          {discount ? (
            <div className="card__price">
              <span>{price.toFixed(1)} </span>
              {(price - (price * discount) / 100).toFixed(1)} BYN
            </div>
          ) : (
            <div className="card__price">{price.toFixed(1)} BYN</div>
          )}

          <div className="card__btns">
            <button className="card__buy">Купить</button>
            <button className="card__cart">
              <img width={24} src={cart} alt="cart" />
            </button>
          </div>
        </div>
        <div className="card__add">
          <img width={24} src={heart} alt="heart" />
          <img width={24} src={bell} alt="bell" />
        </div>
      </div>
    );
  }
}

export default Card;
