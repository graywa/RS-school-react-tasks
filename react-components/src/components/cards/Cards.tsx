import React from 'react';
import Card from '../card/Card';
import './Cards.css';
import { goods } from './data';

function Cards() {
  return (
    <div className="cards">
      {goods.map((el) => {
        return <Card key={el.id} {...el} />;
      })}
    </div>
  );
}

export default Cards;
