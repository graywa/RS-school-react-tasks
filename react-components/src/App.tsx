import React from 'react';
import { Route, Routes } from 'react-router-dom';
import './App.css';
import Header from './components/header/Header';
import AboutUs from './pages/about-us/AboutUs';
import Forms from './pages/forms/Forms';
import HomePage from './pages/home-page/HomePage';
import Page404 from './pages/page-404/Page404';

class App extends React.Component {
  componentDidMount() {
    localStorage.setItem('searchValue', '');
  }

  render() {
    console.log('render app');
    return (
      <div className="app">
        <Header />
        <div className="container">
          <Routes>
            <Route path="/" element={<HomePage />} />
            <Route path="/about" element={<AboutUs />} />
            <Route path="/forms" element={<Forms />} />
            <Route path="*" element={<Page404 />} />
          </Routes>
        </div>
      </div>
    );
  }
}

export default App;
