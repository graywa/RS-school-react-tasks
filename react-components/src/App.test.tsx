import React from 'react';
import { fireEvent, render, screen } from '@testing-library/react';
import { BrowserRouter } from 'react-router-dom';
import Header from './components/header/Header';
import Card from './components/card/Card';
import App from './App';
import SearchBar from './components/search-bar/SearchBar';
import userEvent from '@testing-library/user-event';
import Cards from './components/cards/Cards';
import RickMorty from './pages/rick-morty/RickMorty';
import Page404 from './pages/page-404/Page404';

describe('render', () => {
  test('render page 404', () => {
    const { getByText } = render(<Page404 />);
    const element = getByText(/Page not found/i);
    expect(element).toBeInTheDocument();
  });

  test('render link home', () => {
    const { getByText } = render(
      <BrowserRouter>
        <Header />
      </BrowserRouter>
    );
    const homeLink = getByText(/Rick and Morty/i);
    expect(homeLink).toBeInTheDocument();
  });

  test('render card', () => {
    const { getByText } = render(
      <Card discount={null} photo={''} code={43} price={112} rating={3} title={''} video={true} />
    );
    const price = getByText(/112/i);
    expect(price).toBeInTheDocument();
  });

  test('render all cards', () => {
    const { getAllByText } = render(<Cards />);
    const cards = getAllByText(/купить/i);
    expect(cards.length).toBe(12);
  });

  test('render search input', () => {
    render(
      <BrowserRouter>
        <App />
      </BrowserRouter>
    );
    screen.getByRole('banner');
    expect(screen.getByPlaceholderText('search...'));
  });
});

describe('events', () => {
  test('type in input', () => {
    render(
      <BrowserRouter>
        <App />
      </BrowserRouter>
    );
    fireEvent.change(screen.getByRole('textbox'), { target: { value: 'Hello' } });
    expect(screen.queryByDisplayValue('Hello')).toBeInTheDocument();
  });

  test('more type in input', () => {
    render(
      <BrowserRouter>
        <App />
      </BrowserRouter>
    );
    userEvent.type(screen.getByRole('textbox'), ' World');
    expect(screen.queryByDisplayValue('Hello World')).toBeInTheDocument();
    screen.debug();
  });

  test('click on about link', () => {
    render(
      <BrowserRouter>
        <App />
      </BrowserRouter>
    );
    fireEvent.click(screen.getByText(/About Us/));
    expect(screen.getByRole('heading')).toBeInTheDocument();
  });

  test('focus on input', () => {
    const { getByTestId } = render(<SearchBar status={''} searchValue={''} />);
    const input = getByTestId('search-input');
    expect(input).not.toHaveFocus();
    input.focus();
    expect(input).toHaveFocus();
  });
});

describe('mock localStorage', () => {
  interface EmptyObject {
    [key: string]: string;
  }

  const storageMock = (function () {
    const storage: EmptyObject = {};

    return {
      setItem: function (key: string, value: string) {
        storage[key] = value || '';
      },
      getItem: function (key: string) {
        return storage[key] || null;
      },
      removeItem: function (key: string) {
        delete storage[key];
      },
      get length() {
        return Object.keys(storage).length;
      },
      key: function (i: number) {
        const keys = Object.keys(storage);
        return keys[i] || null;
      },
      clear: function () {
        return {};
      },
    };
  })();

  Object.defineProperty(window, 'localStorage', {
    value: storageMock,
  });

  test('setItem to localStorage', () => {
    const { getByText } = render(
      <BrowserRouter>
        <App />
      </BrowserRouter>
    );
    userEvent.click(getByText(/Rick and Morty/i));
    userEvent.type(screen.getByRole('textbox'), 'No war');
    userEvent.click(getByText(/about us/i));
    expect(localStorage.getItem('searchValue')).toBe('No war');
  });

  test('getItem from localStorage', () => {
    render(
      <BrowserRouter>
        <RickMorty />
      </BrowserRouter>
    );
    expect(screen.getByDisplayValue('No war')).toBeInTheDocument();
  });
});
