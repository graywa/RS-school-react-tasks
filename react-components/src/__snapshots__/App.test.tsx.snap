// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render render as snapshot 1`] = `
<DocumentFragment>
  <header
    class="header"
  >
    <div
      class="container"
    >
      <nav
        class="header__content"
      >
        <a
          aria-current="page"
          class="active"
          href="/"
        >
          Home
        </a>
        <a
          class=""
          href="/about"
        >
          About Us
        </a>
      </nav>
    </div>
  </header>
</DocumentFragment>
`;
